File: sdk/go/schema.go
package sdk

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// fieldEntry describes a single field in the channel schema.
type fieldEntry struct {
	Index uint8  `json:"index"`
	Name  string `json:"name"`
	Type  string `json:"type"`
}

// schemaData is the inner schema payload from the API envelope.
type schemaData struct {
	Fields        []fieldEntry `json:"fields"`
	SchemaVersion uint32       `json:"schema_version"`
}

// schemaEnvelope matches the GreenLab standard API envelope:
// {"success": true, "data": {...}}
type schemaEnvelope struct {
	Success bool       `json:"success"`
	Data    schemaData `json:"data"`
}

// channelSchema caches the resolved schema for a channel.
type channelSchema struct {
	// nameToIndex maps field name → 1-based positional index in the f array.
	// The server returns fields ordered by index, so position i in the f array
	// corresponds to the field with Index == i+1.
	nameToIndex   map[string]uint8
	indexToName   map[uint8]string
	schemaVersion uint32
	// orderedFields holds the fields sorted by index for building the f array.
	orderedFields []fieldEntry
}

// fetchSchema calls GET {baseURL}/v1/channels/{channelID}/schema and builds
// the local schema cache. It sets X-API-Key on the request.
func fetchSchema(ctx context.Context, baseURL, channelID, apiKey string) (channelSchema, error) {
	url := fmt.Sprintf("%s/v1/channels/%s/schema", baseURL, channelID)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return channelSchema{}, fmt.Errorf("sdk: build schema request: %w", err)
	}
	req.Header.Set("X-API-Key", apiKey)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return channelSchema{}, fmt.Errorf("sdk: schema request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return channelSchema{}, fmt.Errorf("sdk: schema fetch returned status %d", resp.StatusCode)
	}

	var env schemaEnvelope
	if err := json.NewDecoder(resp.Body).Decode(&env); err != nil {
		return channelSchema{}, fmt.Errorf("sdk: decode schema response: %w", err)
	}
	if !env.Success {
		return channelSchema{}, fmt.Errorf("sdk: schema response reported failure")
	}

	return buildChannelSchema(env.Data), nil
}

// buildChannelSchema converts the raw API schema into the internal cache.
func buildChannelSchema(data schemaData) channelSchema {
	cs := channelSchema{
		nameToIndex:   make(map[string]uint8, len(data.Fields)),
		indexToName:   make(map[uint8]string, len(data.Fields)),
		schemaVersion: data.SchemaVersion,
		orderedFields: data.Fields,
	}
	for _, f := range data.Fields {
		cs.nameToIndex[f.Name] = f.Index
		cs.indexToName[f.Index] = f.Name
	}
	return cs
}

File: sdk/python/greenlab/schema.py
"""Schema fetching and caching for the GreenLab SDK."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List

import httpx


@dataclass
class FieldEntry:
    index: int
    name: str
    type: str


@dataclass
class ChannelSchema:
    """Cached schema for a channel.

    ``name_to_index`` maps field name → 1-based positional index.
    ``ordered_fields`` lists fields sorted by index (the order they appear
    in the positional ``f`` array sent to the server).
    """

    schema_version: int
    ordered_fields: List[FieldEntry] = field(default_factory=list)
    name_to_index: Dict[str, int] = field(default_factory=dict)
    index_to_name: Dict[int, str] = field(default_factory=dict)


def fetch_schema(base_url: str, channel_id: str, api_key: str) -> ChannelSchema:
    """Fetch the channel schema from the server.

    Calls ``GET {base_url}/v1/channels/{channel_id}/schema`` with the
    ``X-API-Key`` header and parses the response into a :class:`ChannelSchema`.

    Raises :class:`RuntimeError` on non-200 responses or malformed payloads.
    """
    url = f"{base_url}/v1/channels/{channel_id}/schema"
    with httpx.Client() as client:
        resp = client.get(url, headers={"X-API-Key": api_key})

    if resp.status_code != 200:
        raise RuntimeError(f"sdk: schema fetch returned status {resp.status_code}")

    payload = resp.json()
    if not payload.get("success"):
        raise RuntimeError("sdk: schema response reported failure")

    return _build_schema(payload["data"])


def _build_schema(data: dict) -> ChannelSchema:
    """Build a :class:`ChannelSchema` from the raw API ``data`` dict."""
    fields = [
        FieldEntry(index=f["index"], name=f["name"], type=f["type"])
        for f in data.get("fields", [])
    ]
    # Sort by index to guarantee positional array alignment.
    fields.sort(key=lambda f: f.index)

    name_to_index = {f.name: f.index for f in fields}
    index_to_name = {f.index: f.name for f in fields}

    return ChannelSchema(
        schema_version=data.get("schema_version", 0),
        ordered_fields=fields,
        name_to_index=name_to_index,
        index_to_name=index_to_name,
    )
